libgadget/tests/test_density.c

```diff
@@ -0,0 +1,335 @@
+/*Simple test for the tree building functions*/
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include <math.h>
+#include <mpi.h>
+#include <stdio.h>
+#include <time.h>
+#include <gsl/gsl_rng.h>
+
+#include <libgadget/allvars.h>
+#include <libgadget/partmanager.h>
+#include <libgadget/slotsmanager.h>
+#include <libgadget/utils/mymalloc.h>
+#include <libgadget/density.h>
+#include <libgadget/domain.h>
+#include <libgadget/forcetree.h>
+#include <libgadget/timestep.h>
+
+#include "stub.h"
+
+/* The true struct for the state variable*/
+struct forcetree_testdata
+{
+    DomainDecomp ddecomp;
+    gsl_rng * r;
+};
+
+static void set_init_hsml(ForceTree * Tree)
+{
+    int i;
+    #pragma omp parallel for
+    for(i = 0; i < PartManager->NumPart; i++)
+    {
+        int no = force_get_father(i, Tree);
+
+        while(10 * All.DesNumNgb * P[i].Mass > Tree->Nodes[no].u.d.mass)
+        {
+            int p = force_get_father(no, Tree);
+
+            if(p < 0)
+                break;
+
+            no = p;
+        }
+
+        P[i].Hsml =
+            pow(3.0 / (4 * M_PI) * All.DesNumNgb * P[i].Mass / (Tree->Nodes[no].u.d.mass),
+                    1.0 / 3) * Tree->Nodes[no].len;
+    }
+}
+
+/* Perform some simple checks on the densities*/
+static void check_densities(void)
+{
+    int i;
+    double maxHsml=P[0].Hsml, minHsml= P[0].Hsml;
+    #pragma omp parallel for reduction(min:minHsml) reduction(max:maxHsml)
+    for(i=0; i<PartManager->NumPart; i++) {
+        assert_true(isfinite(P[i].Hsml));
+        assert_true(isfinite(SPHP(i).Density));
+        assert_true(SPHP(i).Density > 0);
+        if(P[i].Hsml < minHsml)
+            minHsml = P[i].Hsml;
+        if(P[i].Hsml > maxHsml)
+            maxHsml = P[i].Hsml;
+    }
+    assert_true(isfinite(minHsml));
+    assert_true(minHsml >= All.MinGasHsml);
+    assert_true(maxHsml <= All.BoxSize);
+
+}
+
+static void do_density_test(void ** state, const int numpart, double expectedhsml, double hsmlerr)
+{
+    /*Sort by peano key so this is more realistic*/
```

Comment:
Does it still work without sorting?

Reply:
The sorting is not actually done because it was too much of a pain to keep Slots and P in sync in the test. Sorting only makes the treebuild benchmark more realistic, it does not affect correctness.